base/worker_thread.cpp

```diff
@@ -18,61 +18,129 @@ WorkerThread::~WorkerThread()
 
 bool WorkerThread::Push(Task && t)
 {
-  std::lock_guard<std::mutex> lk(m_mu);
-  if (m_shutdown)
-    return false;
-  m_queue.emplace(std::forward<Task>(t));
-  m_cv.notify_one();
-  return true;
+  return TouchQueues([&]() { m_immediate.emplace(move(t)); });
 }
 
 bool WorkerThread::Push(Task const & t)
 {
-  std::lock_guard<std::mutex> lk(m_mu);
-  if (m_shutdown)
-    return false;
-  m_queue.emplace(t);
-  m_cv.notify_one();
-  return true;
+  return TouchQueues([&]() { m_immediate.emplace(t); });
+}
+
+bool WorkerThread::PushDelayed(Duration const & delay, Task && t)
+{
+  // NOTE: this code depends on the fact that steady_clock is the same
+  // for different threads.
+  auto const when = Now() + delay;
+  return TouchQueues([&]() { m_delayed.emplace(when, move(t)); });
+}
+
+bool WorkerThread::PushDelayed(Duration const & delay, Task const & t)
+{
+  auto const when = Now() + delay;
+  return TouchQueues([&]() { m_delayed.emplace(when, t); });
 }
 
 void WorkerThread::ProcessTasks()
 {
-  queue<Task> pending;
+  ImmediateQueue pendingImmediate;
+  DelayedQueue pendingDelayed;
 
   while (true)
   {
     Task task;
 
     {
       unique_lock<mutex> lk(m_mu);
-      m_cv.wait(lk, [this]() { return m_shutdown || !m_queue.empty(); });
+      if (!m_delayed.empty())
+      {
+        // We need to wait for the moment when the delayed task must
+        // be executed, but may be interrupted earlier, in case of
+        // immediate task or another delayed task that must be
+        // executed earlier.
```

Comment:
Could you please rephrase this sentence?

Reply:
Fixed.